rfcs/0061-reusable-source.md

```diff
@@ -0,0 +1,83 @@
+---
+feature: Reusable Source Executor
+authors:
+  - "Eric Fu"
+start_date: "2023/08/17"
+---
+
+# Reusable Source Executor
+
+
+## Motivation
+
+Conceptually, a source is very different from a table or a sink in terms of whether it is instantiated. A source defined by a `CREATE SOURCE` statement is not an instance, it's just a set of metadata stored in the catalog until it's referenced by a materialized view, at which point a `SourceExecutor` is instantiated.
+
+As a result, when creating multiple materialized views on top of one source, the `SourceExecutor` is not reused between them. It leads to multiple times of resource utilization and multiple consuming offsets, metrics, etc. Furthermore, it becomes a bug if the source definition uses some exclusive property, such as the consumer group of Kafka source.
+
+## Design
+
+Recall that most source has more or less historical data. For message brokers including Kafka, Pulsar, Kinesis, etc, this can be configured with option `scan.startup.mode`. When `scan.startup.mode` is not set to `latest`, the consumer must read the historical data before consuming any new events. This process is implicitly done by the client library, and RisingWave is not aware of it.
+
+Let's assume that the Kafka client library is stupid and RisingWave needs to read the historical data by itself. Then, it turns out this procedure is quite similar to the backfilling procedure of tables. Recall that when a new materialized view is created, we introduce a `BackfillExecutor` to read the historical data. Here we can solve the problem with the same idea by dividing it into 2 phases: 1) backfill and 2) consume new events.
+
+![](images/0061-reusable-source/reusable-source.png)
+
+Here,
+
+- `KafkaSourceExecutor` is created on executing `CREATE SOURCE` statement, although a trivial optimization is doing nothing when there is no downstream operator attached.
```

Comment:
It would be better to clarify the scope of this rfc. Does it mean all `CREATE SOURCE` will be instantiated a corresponding streaming job after this rfc has been done? Which is a fundamental changes to me.

Reply:
>  Does it mean all CREATE SOURCE will be instantiated a corresponding streaming job after this rfc has been done? 

Correct. 

> Which is a fundamental changes to me.

Yes, I agree, but mostly conceptually. The implementation change seems to be acceptable to me.

As stated in the text, currently, the behavior of source is somehow special. So it might be a problem that need to be fixed. I am not sure how users feel about it.

Reply:
+1 to make `CREATE SOURCE` as a streaming job, which is a better way to share same configuration among multiple downstream mv jobs and may save resources. And we don't need to introduce a new `Changelog` stream job proposed in the [RFC: Share a change log stream for multiple CDC tables ](https://www.notion.so/risingwave-labs/RFC-Share-a-change-log-stream-for-multiple-CDC-tables-1f1581d76da2458ba6fafca0c5252226).

Btw I have some discussion with @BugenZhao about the implementation. Most of the work should be done on Meta to add a new kind of `Source` stream job or we can add a dumb sink node behind `Source` to treat it as a `Sink` stream job.